<a href="https://colab.research.google.com/github/antoniosolis4536-collab/MIAAD/blob/main/Actividad_Calificada_07_Datos_Temporales_274706.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |ANTONIO ISMAEL SOLIS MARTINEZ |
| **Matrícula** |274706 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [67]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [68]:
# Escribe tu código aquí
print("Tipo de dato:",taxis['pickup'].dtype)
print("Tipo de dato:",taxis['dropoff'].dtype)

#verificar antes de convertir
nulos_antes = taxis[['pickup', 'dropoff']].isnull().sum()
print("Nulos antes de conversion:\n", nulos_antes)

#hacer la conversion segun como vienen los datos en la tabla
taxis['pickup'] = pd.to_datetime(taxis['pickup'], format='%Y-%m-%d %H:%M:%S')
taxis['dropoff'] = pd.to_datetime(taxis['dropoff'], format='%Y-%m-%d %H:%M:%S')

#verificar nulos (solo para asegurar)
nulos_despues = taxis[['pickup', 'dropoff']].isnull().sum()
print("Nulos despues de conversion:\n", nulos_despues)


Tipo de dato: object
Tipo de dato: object
Nulos antes de conversion:
 pickup     0
dropoff    0
dtype: int64
Nulos despues de conversion:
 pickup     0
dropoff    0
dtype: int64


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [69]:
# Escribe tu código aquí
print("Fecha mínima de pickup:", taxis['pickup'].min())
print("Fecha máxima de pickup:", taxis['pickup'].max())
print("Fecha mínima de dropoff:", taxis['dropoff'].min())
print("Fecha máxima de dropoff:", taxis['dropoff'].max())

fuera_marzo = taxis[~((taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3))]
print("Viajes fuera de marzo de 2019:", len(fuera_marzo))

p_marzo_d_fuera_marzo = taxis[(taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3) &
                              ~((taxis['dropoff'].dt.year == 2019) & (taxis['dropoff'].dt.month == 3))]
print("Viajes con pickup en marzo y dropoff fuera de marzo:", len(p_marzo_d_fuera_marzo))


Fecha mínima de pickup: 2019-02-28 23:29:03
Fecha máxima de pickup: 2019-03-31 23:43:45
Fecha mínima de dropoff: 2019-02-28 23:32:35
Fecha máxima de dropoff: 2019-04-01 00:13:58
Viajes fuera de marzo de 2019: 1
Viajes con pickup en marzo y dropoff fuera de marzo: 1


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [70]:
# Escribe tu código aquí
#Viajes del 10-Marzo
viajes_10Marzo = taxis[taxis['pickup'].dt.date == pd.Timestamp('2019-03-10').date()]
conteo_por_hora = viajes_10Marzo['pickup'].dt.hour.value_counts().sort_index()
print("Viajes por hora:\n", conteo_por_hora)
print("Viajes a las 2:00:", conteo_por_hora.get(2, 0))

pickup_localizado = taxis['pickup'].dt.tz_localize('America/New_York', nonexistent='NaT')
print('Valores NaT despues de localizar American/New_York:', pickup_localizado.isna().sum())

Viajes por hora:
 pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
Name: count, dtype: int64
Viajes a las 2:00: 0
Valores NaT despues de localizar American/New_York: 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

_Escribe aquí tu respuesta._ Pues primero que nada no hubo datos nulos ni antes ni despues de la conversion, solo encontre una fecha fuera de Marzo. Y en lo que respecta al cambio de horario es interesante que no hay datos a esa hora, lo mas seguro es que no fueron registrados, porque al forzar tz_localize no hay registros como si no existiera esa hora.

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [71]:
# Escribe tu código aquí
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60
print(taxis['duracion_min'].describe())

count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [72]:
# Escribe tu código aquí
negativas = (taxis['duracion_min'] < 0).sum()
print("Viajes con duracion negativa:", negativas)

ceros = (taxis['duracion_min'] == 0).sum()
print("Viajes con duracion cero:", ceros)

menor_1min = (taxis['duracion_min'] < 1).sum()
print("Viajes con duracion menor a 1 minuto:", menor_1min)

mayor_3horas = (taxis['duracion_min'] > 180).sum()
print("Viajes con duracion mayor a 3 horas:", mayor_3horas)

taxis[taxis['duracion_min'] == 0][['pickup', 'dropoff', 'distance', 'fare', 'total']]


Viajes con duracion negativa: 0
Viajes con duracion cero: 6
Viajes con duracion menor a 1 minuto: 72
Viajes con duracion mayor a 3 horas: 0


,pickup,dropoff,distance,fare,total
1690,2019-03-22 06:24:14,2019-03-22 06:24:14,0.0,72.0,72.0
5493,2019-03-06 12:14:00,2019-03-06 12:14:00,0.0,10.0,10.0
5624,2019-03-11 14:04:50,2019-03-11 14:04:50,0.0,2.5,3.3
5638,2019-03-13 12:22:00,2019-03-13 12:22:00,0.0,10.0,10.0
6083,2019-03-19 15:34:00,2019-03-19 15:34:00,0.0,5.0,5.0
6356,2019-03-01 16:58:23,2019-03-01 16:58:23,0.0,3.0,4.8


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [73]:
# Escribe tu código aquí
velocidad_calculada = taxis['distance'] / (taxis['duracion_min'] / 60)
taxis['velocidad_mph'] = velocidad_calculada.where(taxis['duracion_min'] > 0)
print("Viajes con velocidad mayor a 60 mph:", (taxis['velocidad_mph'] > 60).sum())

#Viajes con mayor velocidad
taxis.sort_values('velocidad_mph', ascending=False)[['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']].head(5)[
    ['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']]


Viajes con velocidad mayor a 60 mph: 11


,pickup,dropoff,duracion_min,distance,velocidad_mph,fare
770,2019-03-02 03:16:59,2019-03-02 03:17:06,0.116667,9.40,4834.285714,2.5
6089,2019-03-17 23:16:34,2019-03-17 23:16:52,0.300000,10.10,2020.000000,12.0
491,2019-03-07 07:11:33,2019-03-07 07:11:39,0.100000,1.60,960.000000,2.5
5284,2019-03-22 06:24:02,2019-03-22 06:24:11,0.150000,1.70,680.000000,2.5
3241,2019-03-24 17:19:27,2019-03-24 17:19:29,0.033333,0.34,612.000000,65.0


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

In [74]:
taxis[taxis['duracion_min'] < 1][['pickup', 'dropoff', 'distance', 'fare', 'total']]

,pickup,dropoff,distance,fare,total
42,2019-03-30 23:59:14,2019-03-30 23:59:17,0.00,80.0,100.38
88,2019-03-16 05:08:02,2019-03-16 05:08:20,0.07,2.5,3.80
120,2019-03-21 17:21:44,2019-03-21 17:21:49,0.00,2.5,4.30
400,2019-03-09 17:26:02,2019-03-09 17:26:10,0.90,2.5,5.80
458,2019-03-24 01:08:20,2019-03-24 01:08:34,0.02,2.5,6.30
...,...,...,...,...,...
6332,2019-03-19 06:57:36,2019-03-19 06:58:15,0.08,2.5,3.30
6355,2019-03-16 13:24:04,2019-03-16 13:24:13,0.00,2.5,3.30
6356,2019-03-01 16:58:23,2019-03-01 16:58:23,0.00,3.0,4.80
6357,2019-03-19 10:43:58,2019-03-19 10:44:07,0.00,2.5,3.30


**Respuesta 2.4:**

_Escribe aquí tu respuesta._


a) No, no todos los de menos de 1 minuto son error. Algunos sí tienen distancia y tarifa en 0 (esos claramente son cancelados o mal registrados), pero otros tienen tarifa cobrada normal, esos probablemente sí fueron viajes cortos, pero algunos creo no son un error.

b) Yo excluiría: duración menor o igual a 0 (es imposible), duración mayor a 3 horas (ya es un viaje rarísimo para taxi), y velocidad mayor a 60 mph (nadie va a esa velocidad en tráfico de ciudad). También fuera del periodo de marzo 2019, aunque en este caso solo encontré uno.

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [75]:
# Escribe tu código aquí
taxis['hora'] = taxis['pickup'].dt.hour
taxis['dia_num'] = taxis['pickup'].dt.dayofweek
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()
taxis['fin_semana'] = taxis['dia_num'].isin([5, 6])
taxis['fecha'] = taxis['pickup'].dt.date
taxis['cruza_medianoche'] = taxis['fecha'] != taxis['dropoff'].dt.date

print("Viajes que cruzan la medianoche:", taxis['cruza_medianoche'].sum())
taxis.head()

Viajes que cruzan la medianoche: 50


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,...,dropoff_borough,duracion,duracion_min,velocidad_mph,hora,dia_num,dia_nombre,fin_semana,fecha,cruza_medianoche
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,...,Manhattan,0 days 00:06:15,6.250000,15.360000,20,5,Saturday,True,2019-03-23,False
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,...,Manhattan,0 days 00:07:05,7.083333,6.691765,16,0,Monday,False,2019-03-04,False
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,...,Manhattan,0 days 00:07:24,7.400000,11.108108,17,2,Wednesday,False,2019-03-27,False
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,...,Manhattan,0 days 00:25:52,25.866667,17.860825,1,6,Sunday,True,2019-03-10,False
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,...,Manhattan,0 days 00:09:32,9.533333,13.594406,13,5,Saturday,True,2019-03-30,False


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [76]:
# Escribe tu código aquí
patron_hora = taxis.groupby('hora').agg(
    {'duracion_min': ['count', 'median']})

print(patron_hora)
print("Hora con más viajes:", patron_hora['duracion_min']['count'].idxmax())
print("Hora con menos viajes:", patron_hora['duracion_min']['count'].idxmin())


     duracion_min           
            count     median
hora                        
0             205  10.733333
1             110   9.633333
2             101   9.183333
3              67   9.583333
4              57  10.500000
5              51   8.700000
6             142   7.833333
7             221   9.583333
8             313  11.316667
9             319  11.616667
10            325  11.883333
11            295  12.033333
12            334  12.775000
13            316  10.558333
14            360  11.841667
15            330  12.058333
16            336  11.583333
17            388  11.975000
18            417  11.100000
19            406  10.650000
20            368   9.258333
21            355  11.083333
22            321   9.816667
23            296  10.575000
Hora con más viajes: 18
Hora con menos viajes: 5


### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [77]:
# Escribe tu código aquí
patron_dia = taxis.groupby(['dia_num', 'dia_nombre']).agg(
    numero_viajes=('dia_num', 'size'),
    mediana_duracion=('duracion_min', 'median'),
    mediana_tarifa=('fare', 'median')
).reset_index().sort_values('dia_num')

fechas_por_dia = taxis.groupby(['dia_num', 'dia_nombre'])['fecha'].nunique().reset_index(name='fechas_distintas')
patron_dia = patron_dia.merge(fechas_por_dia, on=['dia_num', 'dia_nombre'])
patron_dia['promedio_viajes_por_fecha'] = patron_dia['numero_viajes'] / patron_dia['fechas_distintas']

print(patron_dia)

   dia_num dia_nombre  numero_viajes  mediana_duracion  mediana_tarifa  \
0        0     Monday            708         10.541667             9.5   
1        1    Tuesday            825         11.216667             9.5   
2        2  Wednesday            966         11.291667             9.5   
3        3   Thursday            905         11.766667             9.5   
4        4     Friday           1115         11.150000             9.5   
5        5   Saturday           1046         10.466667             9.0   
6        6     Sunday            868          9.983333             9.0   

   fechas_distintas  promedio_viajes_por_fecha  
0                 4                     177.00  
1                 4                     206.25  
2                 4                     241.50  
3                 5                     181.00  
4                 5                     223.00  
5                 5                     209.20  
6                 5                     173.60  


### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

_Escribe aquí tu respuesta._


a) La hora con mas viajes son las 18 hrs y la de mas larga duracion son las 12 hrs (12.7 min) entonces no coinciden exacto, porque a la hora de más movimiento (tráfico) los viajes tardan más aunque no sean necesariamente la hora con más viajes registrados. Puede ser por el tráfico de la hora pico laboral.

b) Porque marzo no reparte igual los días, no hay el mismo número de lunes (4) que de domingos (5)en ese mes (aunque curiosamente es similar el dato con 177 y 173 respectivamente). Si solo se compara el total, un día que cayó más veces en el mes deberia tener mas viajes aunque en realidad no sea más concurrido por fecha individual, como por ejemplo el miercoles es el mas concurrido aunque solo tiene 4 dias.

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [78]:
# Escribe tu código aquí
condicion_duracion_valida = taxis['duracion_min'] > 0 & (taxis['duracion_min'] <=180)
condicion_velocidad_valida = taxis['velocidad_mph'] <= 60
condicion_periodo_valido = (taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3)

excluidos_por_duracion = (~condicion_duracion_valida).sum()
excluidos_por_velocidad = (~condicion_velocidad_valida).sum()
excluidos_por_periodo = (~condicion_periodo_valido).sum ()

taxis_limpio = taxis[condicion_duracion_valida & condicion_velocidad_valida & condicion_periodo_valido].copy()

print("Viajes excluidos por duracion:", excluidos_por_duracion)
print("Viajes excluidos por velocidad:", excluidos_por_velocidad)
print("Viajes excluidos por periodo:", excluidos_por_periodo)
print('Total_antes:', len(taxis))
print('Total_despues:', len(taxis_limpio))
print('Excluidos_total:', len(taxis) - len(taxis_limpio))

# Verificar
print('Duracion_minima_limpio', taxis_limpio['duracion_min'].min())
print('Duracion_maxima_limpio', taxis_limpio['duracion_min'].max())
print('Velocidad_minima_limpio', taxis_limpio['velocidad_mph'].min())
print('Velocidad_maxima_limpio', taxis_limpio['velocidad_mph'].max())


Viajes excluidos por duracion: 6
Viajes excluidos por velocidad: 17
Viajes excluidos por periodo: 1
Total_antes: 6433
Total_despues: 6415
Excluidos_total: 18
Duracion_minima_limpio 0.05
Duracion_maxima_limpio 107.66666666666667
Velocidad_minima_limpio 0.0
Velocidad_maxima_limpio 54.174757281553404


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [79]:
# Escribe tu código aquí
manana = taxis_limpio[(taxis_limpio['hora'] >= 7) & (taxis_limpio['hora'] <= 9)]

habil = manana[~manana['fin_semana']]
fin_semana = manana[manana['fin_semana']]

for nombre, grupo in [('Dias Habiles', habil), ('Fin de Semana', fin_semana)]:
    print(nombre)
    print('Numero de viajes:', len(grupo))
    print('Mediana de duracion:', grupo['duracion_min'].median())
    print('Mediana de velocidad:', grupo['velocidad_mph'].median())
    print()

Dias Habiles
Numero de viajes: 687
Mediana de duracion: 11.966666666666667
Mediana de velocidad: 9.094736842105263

Fin de Semana
Numero de viajes: 163
Mediana de duracion: 7.633333333333334
Mediana de velocidad: 13.47305389221557



### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

_Escribe aquí tu respuesta._


**Conversión:** usé pd.to_datetime con formato '%Y-%m-%d %H:%M:%S'. Confirmé que no se perdió ningún dato comparando nulos antes y después, salió 0 en ambos.

**Zona horaria:** los datos están en hora local de Nueva York, según el contexto de la actividad. Lo confirmé viendo que no hay viajes a las 2:00 del 10 de marzo, hora que no existió oficialmente ese día por el cambio de horario.

**Criterio de exclusión:** quité duración menor o igual a 0 y mayor a 3 horas, velocidad mayor a 60 mph, y cualquier viaje fuera de marzo 2019 (aunque solo hubo uno). Los umbrales los puse pensando en qué es físicamente razonable para un taxi en ciudad.

**Resultado:** Segun lo que realizo en el punto 4.1 los resultados fueron: datos excluidos por velocidad 11, por duracion 6, y por periodo 1, es decir quitando 18 registros.

**Hallazgo 4.2:** Hablando de este punto pues nos dimos cuenta que la duracion de los viajes son mas altos entre semana (11.96) muy seguramente por el trafico, pero tambien la velocidad es mas lenta, esa puede ser una relacion de porque duran mas. Caso contrario el fin de semana que van mas rapido y la media de duracion es menor(7.63)

**Limitaciones:** los umbrales que usé son criterios razonables pero no absolutos, alguien más podría justificar otros números distintos según el objetivo del análisis.

---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

_Escribe aquí tu respuesta._


**1.** La ventaja de poner el formato explícito es que es más rápido y no deja que pandas "adivine" mal como por ejemplo, con fechas ambiguas tipo 01/03/2025 pandas podría interpretarlo como enero 3 en vez de marzo 1, sin avisarte del error. Ahí es donde la inferencia falla en silencio, sin marcar ningún error, solo te da el dato mal interpretado.

**2.** Si pickup estuviera en UTC y dropoff en hora local, la resta dropoff - pickup daría una duración totalmente incorrecta (de más o de menos varias horas, dependiendo de la diferencia horaria). Lo hubiera detectado igual que detecté el problema del cambio de horario: revisando si hay coherencia en los patrones por hora, o si salen duraciones negativas o rarísimas en cantidad.

**3.** Si no se limpian, el modelo aprendería de datos que físicamente no tienen sentido (viajes de duración negativa, velocidades imposibles), lo que metería ruido y le enseñaría patrones falsos. Esto haría que las predicciones fueran menos confiables para viajes reales.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.